pkg/apis/serving/v1alpha1/inference_graph.go

```diff
@@ -42,6 +42,19 @@ type InferenceGraph struct {
 // InferenceGraphSpec defines the InferenceGraph spec
 // +k8s:openapi-gen=true
 type InferenceGraphSpec struct {
+	/*
+		Example of how to add headers in IG Spec:
+		headers: {
+		 "propagate": [
+			"Custom-Header1",
+			"Custom-Header2"
+		  ]
+		}
+		Note: Making Headers, a map of strings, gives the flexibility to extend it in the future to support adding more
+		operations on headers. For example: Similar to "propagate" operation, one can add "transform" operation if they
+		want to transform headers keys or values before passing down to nodes.
+	*/
+	Headers map[string][]string `json:"headers,omitempty"`
```

Comment:
Not sure if this field is at the right level, do we envision that there are other configurations which need to be overridden on the `InferenceGraph` spec?

Reply:
@yuzisun Yes, there is another PR for adding the affinity spec as well resource spec. https://github.com/kserve/kserve/pull/2711

Reply:
So, we have enabled headers as an optional filed in the spec  in #2396 , but it was only controlled globally. This PR is to let it be overridden if specified at spec level.

Reply:
Affinity and resources are standard pod and container fields which exist on inference service as well, I am a bit concerned about that this can open the door for introducing a set of random configuration fields on inference graph spec.

Reply:
For configuration overrides, we generally use annotations such as `DeploymentMode` in kserve and knative autoscaling annotations.

You can see the knative examples here with global config and annotation overrides
https://knative.dev/docs/serving/autoscaling/scale-to-zero/#scale-to-zero-last-pod-retention-period

Reply:
@navneetraju : Do you want try getting these overrides from annotations as Dan suggested ? Since these are not standard fields for a spec, so does make sense that it will encourage others in future to add more non-standard fields to the spec.

Reply:
Sure, i think it makes sense. My only concern is that we are treating the headers as an object with sub-objects such as propagate, since we want to extensible where we may want to add a block headers config for example. Do we want to represent such objects as an annotation or do we restrict this to propagate only, where give an annotation such as:
```
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: inferencegraph***
spec:
  metadata:
    annotations:
      propagate-headers: "header_a,header_b"
  nodes:
    root:
      routerType: Sequence
      steps:
        - serviceUrl: http://postprocessor.default.svc.cluster.local/v1/models/postprocessor:predict
```

Reply:
@navneetraju : I guess we can do an annotation like this:

`serving.kserve.io/inferencegraph-headers: {"propagate":["Custom-Header1","Custom-Header2"]}`

This annotation will accommodate all rules for propagation/blocking.

@sivanantha321 : Your thoughts ?

Reply:
@rachitchauhan43 What if we introduce new features as you suggested [here](https://github.com/kserve/kserve/blob/430da16356129f0c0f5864d8b29e002a84ca5d55/pkg/controller/v1alpha1/inferencegraph/controller.go#L83). Would those be also accommodated by this annotation or should we introduce annotation just for propagating headers ?

Reply:
@sivanantha321 :  I think let's just keep one annotation that tells what to do with headers. So, if we add a functinality say `drop` certain headers, then it would look like this:
```
serving.kserve.io/inferencegraph-headers: {"propagate":["Custom-Header1","Custom-Header2"],"drop":["Custom-Header3"]}
```

Reply:
@navneetraju : Are you still interested i reviving this PR ? If not , please let us know. We will take over then.